src/metrics/misc.ts

```diff
@@ -207,39 +217,93 @@ export class MetadataMetric implements Metric {
     } else if (result === '[Result "1/2-1/2"]') {
       this.ties++;
     } else {
-      throw new Error('Invalid result');
+      console.log('Invalid result');
     }
 
     // extract openings from metadata
     const opening = metadata?.find((item) => item.startsWith('[Opening "'))
     ?.replace('[Opening "', '')
     ?.replace('"]', '');
 
-    if (opening.toLowerCase() == "bongcloud") {
-      this.bongcloud++;
+    if(opening) {
+      if (opening.toLowerCase() == "bongcloud") {
+        this.bongcloud++;
+      }
+  
+      // add opening to openings object
+      if (this.openings[opening]) {
+        this.openings[opening].appearances++;
+        if (result === '[Result "1-0"]') {
+          this.openings[opening].whiteWins++;
+        } else if (result === '[Result "0-1"]') {
+          this.openings[opening].blackWins++;
+        } else if (result === '[Result "1/2-1/2"]') {
+          this.openings[opening].whiteWins += 0.5;
+          this.openings[opening].blackWins += 0.5;
+          this.openings[opening].ties++;
+        }
+        this.openings[opening].whiteToBlackWinRatio = this.openings[opening].whiteWins / this.openings[opening].blackWins;
+      } else {
+        this.openings[opening] = {
+          appearances: 1,
+          blackWins: result === '[Result "0-1"]' ? 1 : (result === '[Result "1/2-1/2"]' ? 0.5 : 0),
+          whiteWins: result === '[Result "1-0"]' ? 1 : (result === '[Result "1/2-1/2"]' ? 0.5 : 0),
+          ties: result === '[Result "1/2-1/2"]' ? 1 : 0,
+          whiteToBlackWinRatio: result === '[Result "1-0"]' ? 1 : 0,
+        };
+      }
     }
 
-    // add opening to openings object
-    if (this.openings[opening]) {
-      this.openings[opening].appearances++;
-      if (result === '[Result "1-0"]') {
-        this.openings[opening].whiteWins++;
-      } else if (result === '[Result "0-1"]') {
-        this.openings[opening].blackWins++;
+
+    // identify game endings
+    let gameEnd = metadata.find(item => item.startsWith('[Termination'));
+    const lastMove = game[game.length - 1].move;
+
+    if (gameEnd === '[Termination "Normal"]') {
+      if (lastMove.originalString.includes('#') || this.chess.isCheckmate()) {
+        this.gameEndings['checkmate'] = (this.gameEndings['checkmate'] || 0) + 1;
       } else if (result === '[Result "1/2-1/2"]') {
-        this.openings[opening].whiteWins += 0.5;
-        this.openings[opening].blackWins += 0.5;
-        this.openings[opening].ties++;
+        this.gameEndings['draw'] = (this.gameEndings['draw'] || 0) + 1;
+        if (this.chess.isStalemate()) {
+          this.gameEndings['stalemate'] = (this.gameEndings['stalemate'] || 0) + 1;
+        } else if (this.chess.isInsufficientMaterial()) {
+          this.gameEndings['insufficient material'] = (this.gameEndings['insufficient material'] || 0) + 1;
+        } 
+      } else {
```

Comment:
might note that this is an implicit assumption that metadata is valid

Reply:
what do you mean? this is buried under multiple if statements that prob ensure metadata validity